### PR TITLE
FEATURE: Add BA3031.EnableClangSafeStack

### PR DESCRIPTION
**Pull request by shaopeng-gh**
This new rule applies to Clang only since GCC does not support it.
More info:
https://clang.llvm.org/docs/SafeStack.html
https://github.com/microsoft/binskim/issues/634

Also fixed:
1. rename BA3030.UseCheckedFunctionsWithGcc to BA3030.UseGccCheckedFunctions
2. rename MetadataConditions.ElfIsCoreNoneOrObject to MetadataConditions.ElfIsCoreNoneOrRelocatable

### CODE REVIEW

File: src/Test.FunctionalTests.BinSkim.Driver/BaselineTestData/Expected/Native_x64_RustC_Rust_debuginfo2_v1.58.1.exe.sarif [hunk 1]
```diff
@@ -951,7 +951,7 @@
             },
             {
               "id": "BA2026",
-              "name": "EnableAdditionalSdlSecurityChecks",
+              "name": "EnableMicrosoftCompilerSdlSwitch",
```
Comment (shaopeng-gh) [inline src/Test.FunctionalTests.BinSkim.Driver/BaselineTestData/Expected/Native_x64_RustC_Rust_debuginfo2_v1.58.1.exe.sarif:954]: >EnableMicrosoftCompilerSdlSwitch [](http://example.com/codeflow?start=23&length=32)

This is not related to the current change, fix by the way.

File: src/Test.FunctionalTests.BinSkim.Rules/RuleTests.cs [hunk 2]
```diff
@@ -1416,6 +1416,24 @@ public void BA3030_UseCheckedFunctionsWithGCC_NotApplicable()
             this.VerifyApplicability(new UseCheckedFunctionsWithGcc(), new HashSet<string>());
         }
 
+        [Fact]
+        public void BA3031_EnableSafeStackWithClang_Pass()
```
Comment (shaopeng-gh) [inline src/Test.FunctionalTests.BinSkim.Rules/RuleTests.cs:1420]: >public void BA3031_EnableSafeStackWithClang_Pass() [](http://example.com/codeflow?start=8&length=50)

the combination of the test files are:
1. version, from latest 14 down to 7
2. language, C and C++

(By these tests we found that lower version has different symbol name)

File: src/Test.FunctionalTests.BinSkim.Driver/BaselineTestData/Expected/Native_x64_RustC_Rust_debuginfo2_v1.58.1.exe.sarif (hunk 1 above)
Comment (marmegh) [reply src/Test.FunctionalTests.BinSkim.Driver/BaselineTestData/Expected/Native_x64_RustC_Rust_debuginfo2_v1.58.1.exe.sarif:954]: Nice. Thanks for fixing this.

File: src/Test.FunctionalTests.BinSkim.Rules/RuleTests.cs (hunk 2 above)
Comment (marmegh) [reply src/Test.FunctionalTests.BinSkim.Rules/RuleTests.cs:1420]: Awesome, do we have build script examples to include for these?

File: docs/FunctionalTestBuildScripts.md [hunk 3]
```diff
@@ -4,6 +4,52 @@ This file records scripts used to compile the test files, in alphabetical order.
 Base scenario is a simple hello world program built with different parameters for testing purpose.
 Test files are located in [BaselineTestData](https://github.com/microsoft/binskim/tree/main/src/Test.FunctionalTests.BinSkim.Driver/BaselineTestData) and [FunctionalTestData](https://github.com/microsoft/binskim/tree/main/src/Test.FunctionalTests.BinSkim.Rules/FunctionalTestData).
 
+## ARM64_CETShadowStack_NotApplicable.exe
```
Comment (shaopeng-gh) [inline docs/FunctionalTestBuildScripts.md:7]: this is just a re-order alphabetically

File: src/Test.FunctionalTests.BinSkim.Rules/RuleTests.cs (hunk 2 above)
Comment (shaopeng-gh) [reply src/Test.FunctionalTests.BinSkim.Rules/RuleTests.cs:1420]: yes! added.

File: docs/FunctionalTestBuildScripts.md [hunk 4]
```diff
@@ -4,6 +4,52 @@ This file records scripts used to compile the test files, in alphabetical order.
 Base scenario is a simple hello world program built with different parameters for testing purpose.
 Test files are located in [BaselineTestData](https://github.com/microsoft/binskim/tree/main/src/Test.FunctionalTests.BinSkim.Driver/BaselineTestData) and [FunctionalTestData](https://github.com/microsoft/binskim/tree/main/src/Test.FunctionalTests.BinSkim.Rules/FunctionalTestData).
 
+## ARM_CETShadowStack_NotApplicable.exe
+
+A simple C++ hellow world program, cross compiled using CMake with the `cl.exe` compiler and `Ninja` generator.
+`CMakePresets.json` should be configured with a `configurePresets` as below:
+
+```json
```
Comment (shaopeng-gh) [inline docs/FunctionalTestBuildScripts.md:12]: >json [](http://example.com/codeflow?start=3&length=4)

I added the language for ``` per linter request. and empty lines.

File: docs/BinSkimRules.md [hunk 5]
```diff
@@ -226,6 +226,28 @@ No checked functions are present/used when compiling '{0}', and it was compiled
 
 ---
 
+## Rule `BA3031.EnableSafeStackWithClang`
```
Comment (michaelcfanning) [inline docs/BinSkimRules.md:229]: >EnableSafeStackWithClang [](http://example.com/codeflow?start=16&length=24)

EnableClangSafeStack?

I will think about the name and help refine this help text a bit further.

File: src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+                    nameof(RuleResources.NotApplicable_InvalidMetadata)
+                };
+
+        public override AnalysisApplicability CanAnalyzeElf(ElfBinary target, Sarif.PropertiesDictionary policy, out string reasonForNotAnalyzing)
+        {
+            IELF elf = target.ELF;
+
+            if (elf.Type == FileType.Core || elf.Type == FileType.None || elf.Type == FileType.Relocatable)
+            {
+                reasonForNotAnalyzing = MetadataConditions.ElfIsCoreNoneOrObject;
```
Comment (michaelcfanning) [inline src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs:50]: >ElfIsCoreNoneOrObject [](http://example.com/codeflow?start=59&length=21)

I would tend to name this 'relocatable' as it is, I think, the most precise *nix term for what we're looking at.

File: src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+                return AnalysisApplicability.NotApplicableToSpecifiedTarget;
+            }
+
+            if (!target.Compilers.Any(c => c.Compiler == ElfCompilerType.Clang && c.Version.Major >= 7))
+            {
+                reasonForNotAnalyzing = MetadataConditions.ElfNotBuiltWithClangV7OrLater;
```
Comment (michaelcfanning) [inline src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs:56]: >ElfNotBuiltWithClangV7OrLater [](http://example.com/codeflow?start=59&length=29)

Use a generic message, BinaryCompiledWithOutdatedTools or just use the existing one ImageCompiledWithOutdatedTools?

File: src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        /// Checks if SafeStack is enabled by Symbols.
+        /// </summary>
+        public override void Analyze(BinaryAnalyzerContext context)
+        {
+            IELF elf = context.ElfBinary().ELF;
+
+            IEnumerable<ISymbolEntry> symbols =
+                ElfUtility.GetAllSymbols(elf).Where(sym => sym.Type == SymbolType.File);
+
+            if (symbols.Any(s => symbolSafeStack.Contains(s.Name)))
```
Comment (michaelcfanning) [inline src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs:74]: Foreach over the symbols, to avoid a 2x traversal in the pathological case. Fire a result and exit if you meet the first condition. Otherwise fall out and fire the pass result.

File: src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+            {
+                reasonForNotAnalyzing = MetadataConditions.ElfIsCoreNoneOrObject;
+                return AnalysisApplicability.NotApplicableToSpecifiedTarget;
+            }
+
+            if (!target.Compilers.Any(c => c.Compiler == ElfCompilerType.Clang && c.Version.Major >= 7))
+            {
+                reasonForNotAnalyzing = MetadataConditions.ElfNotBuiltWithClangV7OrLater;
+                return AnalysisApplicability.NotApplicableToSpecifiedTarget;
+            }
```
Comment (michaelcfanning) [inline src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs:58]: Remove this.

File: src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+                ElfUtility.GetAllSymbols(elf).Where(sym => sym.Type == SymbolType.File);
+
+            if (symbols.Any(s => symbolSafeStack.Contains(s.Name)))
+            {
+                context.Logger.Log(this,
+                       RuleUtilities.BuildResult(ResultKind.Pass, context, null,
+                           nameof(RuleResources.BA3031_Pass),
+                           context.TargetUri.GetFileName()));
+            }
+            else
```
Comment (michaelcfanning) [inline src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs:81]: >else [](http://example.com/codeflow?start=12&length=4)

When you exit the function, you could do a version check, and if it's v3.8 or whatever minimal version we know about, just fire an error.

Otherwise fire a special error that says you should enable SafeStack, you might need to update your version of Clang to do this.

**Review by michaelcfanning — APPROVED** [review]
:shipit:

File: docs/BinSkimRules.md (hunk 5 above)
Comment (shaopeng-gh) [reply docs/BinSkimRules.md]: fixed

File: src/BinSkim.Rules/RuleIds.cs [hunk 11]
```diff
@@ -61,6 +61,7 @@ internal static class RuleIds
         // BA3012-3029 -- saved for future non-compiler/language specific checks.
         // Compiler/Language specific checks follow.
         public const string UseCheckedFunctionsWithGcc = "BA3030";
```
[the remark this answers is not in the document]
Comment (shaopeng-gh) [reply src/BinSkim.Rules/RuleIds.cs]: fixed

File: src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs (hunk 6 above)
Comment (shaopeng-gh) [reply src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs]: fixed to ElfIsCoreNoneOrRelocatable

File: src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs (hunk 7 above)
Comment (shaopeng-gh) [reply src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs]: the comment is replaced by later discussions to fire a special error that says you should enable SafeStack, you might need to update your version of Clang to do this.

File: src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs (hunk 9 above)
Comment (shaopeng-gh) [reply src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs]: Removed the version check only, we still need to Clang check, if people just use Gcc we will skip this image as NotApplicable.

File: src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs (hunk 10 above)
Comment (shaopeng-gh) [reply src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs]: agree with this approach, fixed.

File: src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs (hunk 8 above)
Comment (shaopeng-gh) [reply src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs]: Changed to not use Linq and just use foreach.
(It may not result in 2x traversal though)

File: src/Test.FunctionalTests.BinSkim.Driver/BaselineTestData/Expected/BinSkim.win-x64.ni.dll.sarif.temp2 [hunk 12]
```diff
@@ -1,133 +0,0 @@
-{
```
Comment (shaopeng-gh) [inline src/Test.FunctionalTests.BinSkim.Driver/BaselineTestData/Expected/BinSkim.win-x64.ni.dll.sarif.temp2:1]: remove not needed checked in temp file.
